### PR TITLE
PNV contributions round 2

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
With many thanks to: 
 - Cuong
 - Lai
 - Mai
 - Suu
 - Am
 - Phuc
 - Vi
 - Tuan

Students in second year at Passerelle Numériques Vietnam

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java [hunk 1]
```diff
@@ -120,28 +135,72 @@ public String getMailetInfo() {
...
+               addresses.addAll(getMailAddressesFromHeaderLine(header));
+            }
+        }
+        return addresses.build();
+    }
+
+    private ImmutableList<MailAddress> getMailAddressesFromHeaderLine(String header) throws MessagingException {
+        String unfoldedDecodedString = sanitizeHeaderString(header);
+        Iterable<String> headerParts = Splitter.on(",").split(unfoldedDecodedString);
+        ImmutableList.Builder<MailAddress> result = ImmutableList.builder();
```
Comment (quynhn) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: can we extract this part of code to function?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java [hunk 2]
```diff
@@ -120,28 +135,72 @@ public String getMailetInfo() {
...
+        return result.build();
+    }
+
+    private Collection<MailAddress> readMailAddresses(String headerPart) throws AddressException {
+        AddressList addressList = LenientAddressParser.DEFAULT
+            .parseAddressList(MimeUtil.unfold(headerPart));
+
+        ImmutableList.Builder<Mailbox> mailboxList = ImmutableList.builder();
+
+        for (Address address: addressList) {
```
Comment (quynhn) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: Can we use Guava?

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java [hunk 3]
```diff
@@ -120,28 +135,72 @@ public String getMailetInfo() {
...
+            mailboxList.addAll(convertAddressToMailboxCollection(address));
+        }
+
+        return FluentIterable.from(mailboxList.build())
+            .transform(new Function<Mailbox, MailAddress>() {
```
Comment (quynhn) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: to constant I think

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    public void setUp() throws Exception {
+        testee = new UseHeaderRecipients();
+        mailetContext = FakeMailContext.defaultContext();
+        testee.init(FakeMailetConfig.builder().mailetContext(mailetContext).build());
+    }
+
+    @Test
+    public void serviceShouldSetMimeMessageRecipients() throws Exception {
+
+        String RCPT_1 = "abc1@apache1.org";
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: extract to constant for all kind of those values

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                .addToRecipient(RCPT_1)
+                .addCcRecipient(RCPT_2)
+                .addBccRecipient(RCPT_3)
+                .build())
+            .build();
+
+        testee.service(fakeMail);
+
+        assertThat(fakeMail.getRecipients())
+            .containsOnly(new MailAddress(RCPT_1), new MailAddress(RCPT_2), new MailAddress(RCPT_3));
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: extract "new MailAddress(RCPT_1)" to constant object then we can reduce line of code

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+        assertThat(mailetContext.getSentMails())
```
Comment (quynhn) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: Could we accept multiple assert on the same function?

File: mailet/standard/src/test/java/org/apache/james/transport/matchers/TooManyRecipientsTest.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private TooManyRecipients testee;
+
+    @Before
+    public void setUp() {
+        testee = new TooManyRecipients();
+    }
+
+    @Test (expected = MessagingException.class)
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/matchers/TooManyRecipientsTest.java]: Arf, I did not teach them ExpectedException as it requires @Rule which is harder to teach. Will do the port.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java (hunk 1 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: If you want

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java (hunk 2 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: Nope. Error exception would then go messy.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java (hunk 3 above)
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/UseHeaderRecipients.java]: If you want

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java (hunk 5 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: To a field => error handling

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java (hunk 6 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: The first one is a mistake

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java (hunk 5 above)
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: ok but RCP_* should be constants (BTW you written them in caps :))

**Comment by chibenwa** [comment]
test this please

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            .build();
+
+        testee.service(fakeMail);
+
+        assertThat(fakeMail.getRecipients())
+            .isEmpty();
+    }
+
+    @Test
+    public void serviceShouldGoshEmail() throws Exception {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: Gosh? :)

Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: missing when in the test name

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+            .build();
+
+        testee.service(fakeMail);
+
+        assertThat(fakeMail.getState())
+            .isEqualTo(Mail.GHOST);
+    }
+
+    @Test
+    public void serviceShouldResendTheEmail() throws Exception {
```
Comment (rouazana) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: missing when in the test name

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java (hunk 8 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: No when missing as service always ghost the email

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java (hunk 9 above)
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/UseHeaderRecipientsTest.java]: No when missing as service always resend the email

**Comment by chibenwa** [comment]
Just merged

**Pull request closed without merging**